### ISSUE TITLE
Proposal: decimal floating-point types

### ISSUE THREAD

**Issue by SaintCharlesCode** [ziglang/zig#4221]
What are your thoughts on Zig having decimal floating-point types built into the language? http://speleotrove.com/decimal/
They could have types df32 df64 df128 or something like that.

**Comment by JesseRMeyer** [comment]
There's also the general possibility of user-space defined Fixed Point storage.

**Comment by shawnl** [comment]
you should probably bring this up with llvm. This is a bunch of work (it needs a full soft float library) and that work would probably belong in llvm.

**Comment by SaintCharlesCode** [comment]
That's true. Also clang does not have "(_Decimal32 and friends)" according to the manual https://clang.llvm.org/docs/UsersManual.html It is probably too much of a niche feature.

**Comment by DanB91** [comment]
Came across this article and adding this in could maybe help in places where COBOL is used: https://medium.com/the-technical-archaeologist/is-cobol-holding-you-hostage-with-math-5498c0eb428b

**Comment by matu3ba** [comment]
This is identical to the more poorly named #1974.
Look especially at this nice comment, which elaborates quite alot: https://github.com/ziglang/zig/issues/1974#issuecomment-620336282

Though I disagree that a separate type should be reserved for it, since the type consists of a tuple.

**Comment by zzyxyzz** [comment]
At the moment, #1974 is only about *binary* fixed-point numbers, so it wouldn't help with this issue.

**Comment by matu3ba** [comment]
> At the moment, #1974 is only about _binary_ fixed-point numbers, so it wouldn't help with this issue.

They are very closely related: https://www.tutorialspoint.com/fixed-point-and-floating-point-number-representations,
so I dont see a special reason to separate them.
Since we have comptime, they could probably even be done together.

If you want to be fancy you could also use [binary-coded decimals](https://en.wikipedia.org/wiki/Binary-coded_decimal).

**Comment by zzyxyzz** [comment]
@matu3ba,
Fixed vs floating point and binary vs decimal are orthogonal problems, as I understand them. The former trades range for uniformity of representable numbers (and sometimes also performance). The latter determines which fractional numbers are representable *exactly* and which aren't. E.g., `0.1` is not representable as a binary fraction, regardless of whether it is fixed or floating. Although the rounding errors are usually small, they sometimes do matter in financial and accounting software. And they can compound quickly, with the classic example being:
```zig
var x = 0.01; // type omitted
var i: i32 = 0;
while (i < 12) : (i += 1) {
    x = 1.0 - 99 * x;
}
```
At the end of the 12 iterations, `x` would have the value 0.01 if it is represented as a decimal fraction. But with a binary fraction, each step eats about 6 bits of precision. If `x: f64`, then the result is -79522. This is the kind of thing people are usually trying to prevent when calling for decimal number support.

To be clear, I'm highly skeptical of introducing native decimal fraction support in Zig. I'm just saying that if we do, then #1974 is not sufficient, unless it is extended to specify the base as well.

**Comment by haberman** [comment]
It looks like decimal floating point is an optional feature in C23: https://www.iso.org/standard/68882.html

GCC seems to have some amount of support for it: https://gcc.gnu.org/onlinedocs/gcc/Decimal-Float.html

Example usage: https://godbolt.org/z/d6EzqcrE6

There is talk of implementing this in Clang, but nothing tangible yet AFAICS: https://discourse.llvm.org/t/rfc-decimal-floating-point-support-iso-iec-ts-18661-2-and-c23/62152/1
